code/modules/mob/living/silicon/ai/ai_programs.dm

```diff
@@ -0,0 +1,1017 @@
+	selection_deactivated_message = "<span class='notice'>You rescind the order.</span>"
+
+/datum/spell/ai_spell/ranged/repair_nanites/cast(list/targets, mob/user)
+	var/target = targets[1]
+	if(!check_camera_vision(user, target))
+		revert_cast()
+		return
+	if(!istype(target, /mob/living/silicon/robot) && !istype(target, /obj/machinery/power/apc) && !istype(target, /obj/mecha) && !istype(target, /mob/living/carbon/human/machine))
+		to_chat(user, "<span class='warning'>You can only recharge borgs, mechs, and APCs!</span>")
+		revert_cast()
+		return
+	var/mob/living/silicon/ai/AI = user
+	if(istype(target, /mob/living/carbon/human/machine)  && !AI.universal_adapter)
+		to_chat(user, "<span class='warning'>This software lacks the required upgrade to recharge IPCs!</span>")
+		revert_cast()
+		return
+	if(istype(target, /obj/mecha)|| istype(target, /obj/machinery/power/apc))
+		var/obj/T = target
+		T.obj_integrity += min(T.max_integrity, T.max_integrity * (0.2 +  min(0.3, (0.1 * spell_level))))
+	if(istype(target, /mob/living/silicon/robot))
+		var/mob/living/silicon/robot/T = target
+		var/damage_healed = 20 + (min(30, (10 * spell_level)))
+		T.heal_overall_damage(damage_healed, damage_healed)
+	if(istype(target, /mob/living/carbon/human/machine))
+		var/mob/living/carbon/human/machine/T = target
+		var/damage_healed = AI.adapter_efficiency * (20 + (min(30, (10 * spell_level))))
+		T.heal_overall_damage(damage_healed, damage_healed, TRUE, 0, 1)
+	AI.program_picker.nanites -= 75
+	AI.play_sound_remote(target, 'sound/goonstation/misc/fuse.ogg', 50)
+	camera_beam(target, "medbeam", 'icons/effects/beam.dmi', 10)
+
+/datum/spell/ai_spell/ranged/repair_nanites/on_purchase_upgrade()
+	cooldown_handler.recharge_duration = max(min(base_cooldown, base_cooldown - (max(spell_level - 3, 0) * 60)), cooldown_min)
+
+// Universal Adapter - Unlocks usage of repair nanites and power shunt for IPCs
+/datum/ai_program/universal_adapter
+	program_name = "Universal Adapter"
+	program_id = "universal_adapter"
+	description = "Upgraded firmware allows IPCs to be valid targets for Power Shunt and Repair Nanites, at half efficiency."
+	nanite_cost = 0
+	unlock_text = "Universal adapter installation complete!"
+	upgrade = TRUE
+
+/datum/ai_program/universal_adapter/upgrade(mob/living/silicon/ai/user, first_install = FALSE)
+	..()
+	if(!istype(user))
+		return
+	user.universal_adapter = TRUE
+	user.adapter_efficiency = 0.5 + (0.1 * upgrade_level)
+	installed = TRUE
+
+/datum/ai_program/universal_adapter/downgrade(mob/living/silicon/ai/user)
+	..()
+	if(!istype(user))
+		return
+	user.adapter_efficiency = 0.5 + (0.1 * upgrade_level)
+
+/datum/ai_program/universal_adapter/uninstall(mob/living/silicon/ai/user)
+	..()
+	if(!istype(user))
+		return
+	user.adapter_efficiency = 0.5
+	user.universal_adapter = FALSE
+
+// Door Override - Repairs door wires if the AI wire is not cut
+/datum/ai_program/door_override
+	program_name = "Door Override"
```

Comment:
"Door Override" seems like an awkward name for what it's actually doing. maybe "Airlock Mend" or "Airlock Restore"?